### PR TITLE
Master product api

### PR DESCRIPTION
**Pull request by cp-ps**

### CODE REVIEW

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/upsertProduct.feature [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      | ALBERTA    | 345        | bpartner              |
+      | ALBERTA    | 456        | bpartner              |
+
+    Given the user adds bpartner product
+      | ProductCode  | bpartnerIdentifier | OPT.isActive | OPT.seqNo | OPT.ProductNo | OPT.Description | OPT.EAN  | OPT.GTIN  | OPT.customerLabelName | OPT.ingredients |
+      | code345      | ext-ALBERTA-345    | true         | 10        | test          | test            | ean_test | gtin_test | test                  | test            |
+      | code345      | ext-ALBERTA-456    | true         | 10        | test          | test            | ean_test | gtin_test | test                  | test            |
+
+    Given the user adds product
+      | productIdentifier | ProductCode | Name         | Type | UOMCode | OPT.EAN  | OPT.GTIN  | OPT.Description  | OPT.isActive |
```
Comment (metas-ts) [inline backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/upsertProduct.feature:25]: if these are json-properties, then use the exact property names like `uomCode`

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/upsertProduct.feature [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      | ext-ALBERTA-345   | code345     | Product_Test | ITEM | PCE     | ean_test | gtin_test | test_description | true         |
+
+    And the request is set in context of syncAdvise 'CREATE_OR_MERGE'
```
Comment (metas-ts) [inline backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/upsertProduct.feature:28]:
```suggestion
    And we create a JsonRequestProductUpsert, set syncAdvise to 'CREATE_OR_MERGE', add the JsonRequestProductUpsertItems and store it in the test context
```

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/upsertProduct.feature [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    And no product with value 'code345' exists
+
+  Scenario: create Product request
+
+    Given the user adds external reference
+      | systemname | externalId | externalReferenceType |
+      | ALBERTA    | 345        | bpartner              |
+      | ALBERTA    | 456        | bpartner              |
+
+    Given the user adds bpartner product
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/upsertProduct.feature:19]: The external references are not parts of the payload, changed to "metasfresh contains S_ExternalReferences"

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/product/ProductRestController.java [hunk 4]
```diff
@@ -53,6 +56,13 @@ public ProductRestController(final ProductRestService productRestService)
 		this.productRestService = productRestService;
 	}
 
+	@ApiOperation("Create or update products and corresponding Bpartner-products.")
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/product/ProductRestController.java:59]: please consolidate this class with `ProductsRestController` and add the API-Docu there.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/product/ProductRestController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+{
+	private final ProductRestService productRestService;
+
+	public ProductRestController(final ProductRestService productRestService)
+	{
+		this.productRestService = productRestService;
+	}
+
+	@PutMapping("{orgCode}")
+	public ResponseEntity<JsonResponseUpsert> upsertProducts(@RequestBody @NonNull final JsonRequestProductUpsert request,
```
[the remark this answers is not in the document]
Comment (metas-ts) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/product/ProductRestController.java:67]: also add an `@ApiParam` (=> orientate on `BpartnerRelationRestController.retrieveBPartner()`)

File: backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/RESTUtil.java [hunk 6]
```diff
@@ -148,7 +149,14 @@ public static SyncAdvise mapSyncAdvise(@NonNull final String syncAdvise)
 			case "READ_ONLY":
 				return SyncAdvise.READ_ONLY;
 			default:
-				return null;
+				if (EmptyUtil.isBlank(syncAdvise))
```
Comment (metas-ts) [inline backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/RESTUtil.java:152]: please avoid such deep nesting by moving this if-block to the start of the method

File: backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/product/UpsertProduct_StepDef.java [hunk 7]
```diff
@@ -220,9 +220,9 @@ public void theUserAddsBpartnerExternalReference(@NonNull final DataTable dataTa
 				.map(bPartner -> JsonMetasfreshId.of(bPartner.getC_BPartner_ID())).collect(Collectors.toList());
 
 		dataTableEntries.forEach(dataTableEntry -> {
-			final String externalSystemName = DataTableUtil.extractStringForColumnName(dataTableEntry, "systemname");
-			final String externalId = DataTableUtil.extractStringForColumnName(dataTableEntry, "externalId");
-			final IExternalReferenceType externalReferenceType = getExternalReferenceType(DataTableUtil.extractStringForColumnName(dataTableEntry, "externalReferenceType"));
+			final String externalSystemName = DataTableUtil.extractStringForColumnName(dataTableEntry, "ExternalSystem.Code");
+			final String externalId = DataTableUtil.extractStringForColumnName(dataTableEntry, "ExternalReference");
+			final IExternalReferenceType externalReferenceType = getExternalReferenceType(DataTableUtil.extractStringForColumnName(dataTableEntry, "ExternalReference.Code"));
```
Comment (metas-ts) [inline backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/product/UpsertProduct_StepDef.java:225]:
```suggestion
			final IExternalReferenceType externalReferenceType = getExternalReferenceType(DataTableUtil.extractStringForColumnName(dataTableEntry, "ExternalReferenceType.Code"));
```

**Review by metas-ts — CHANGES_REQUESTED** [review]

**Review by metas-ts — APPROVED** [review]

**Pull request merged**